core/src/test/java/io/parsingdata/metal/expression/value/ElvisExpressionTest.java

```diff
@@ -50,27 +44,78 @@
-        assertThat(eval.get().asNumeric().intValue(), is(equalTo(1)));
+        assertNotNull(eval);
+        assertEquals(1, eval.size);
+        assertThat(eval.head.get().asNumeric().intValue(), is(equalTo(1)));
     }
 
     @Test
     public void elvisRight() throws IOException {
         final ParseResult result = choice.parse(stream(2), enc());
-        final OptionalValue eval = elvisExpression.eval(result.getEnvironment(), enc());
+        final OptionalValueList eval = elvisExpression.eval(result.getEnvironment(), enc());
 
-        assertTrue(eval.isPresent());
-        assertThat(eval.get().asNumeric().intValue(), is(equalTo(2)));
+        assertNotNull(eval);
+        assertEquals(1, eval.size);
+        assertThat(eval.head.get().asNumeric().intValue(), is(equalTo(2)));
     }
 
     @Test
     public void elvisNone() throws IOException {
         final ParseResult result = choice.parse(stream(3), enc());
-        final OptionalValue eval = elvisExpression.eval(result.getEnvironment(), enc());
+        final OptionalValueList eval = elvisExpression.eval(result.getEnvironment(), enc());
 
-        assertFalse(eval.isPresent());
+        assertNotNull(eval);
+        assertTrue(eval.isEmpty());
+    }
+
+    private Token createListToken(final Token left, final Token right) {
+        return
+            seq(left,
+                right
+            );
+    }
+
+    @Test
+    public void elvisList() throws IOException {
+        final ParseResult result = createListToken(seq(any("a"), any("a")), seq(any("b"), any("b"))).parse(stream(1, 2, 3, 4), enc());
+        assertTrue(result.succeeded());
+        final ValueExpression elvis = elvis(ref("a"), ref("b"));
+        final OptionalValueList eval = elvis.eval(result.getEnvironment(), enc());
+        assertEquals(2, eval.size);
+        assertEquals(2, eval.head.get().asNumeric().intValue());
+        assertEquals(1, eval.tail.head.get().asNumeric().intValue());
+    }
+
+    @Test
+    public void elvisListWithEmpty() throws IOException {
+        final ParseResult result = createListToken(seq(any("a"), any("a")), seq(any("b"), any("b"))).parse(stream(1, 2, 3, 4), enc());
+        assertTrue(result.succeeded());
+        final ValueExpression elvis = elvis(ref("c"), ref("b"));
+        final OptionalValueList eval = elvis.eval(result.getEnvironment(), enc());
+        assertEquals(2, eval.size);
+        assertEquals(4, eval.head.get().asNumeric().intValue());
+        assertEquals(3, eval.tail.head.get().asNumeric().intValue());
+    }
+
+    @Test
+    public void elvisListDifferentLengths() throws IOException {
+        final ParseResult result = createListToken(seq(any("a"), any("a")), seq(any("b"), any("b"), any("b"))).parse(stream(1, 2, 3, 4, 5), enc());
+        assertTrue(result.succeeded());
+        final ValueExpression elvis = elvis(ref("a"), ref("b"));
+        final OptionalValueList eval = elvis.eval(result.getEnvironment(), enc());
```

Comment:
Let me work this out:

- `ref("a") = [2,1]`
- `ref("b") = [5,4,3]`
- `elvis(ref("a"), ref("b")) = [2,1,3]`

Got it. :-)